daceml/autodiff/pytorch.py

```diff
@@ -93,119 +94,33 @@ def make_backward_function(model: ONNXModel,
             bwd_copy_state.add_edge(bwd_copy_state.add_read(bwd_arr_name),
                                     None, bwd_copy_state.add_write(name), None,
                                     dace.Memlet(name + "[0]"))
-            replaced_scalars[name] = fwd_arr_name
+            replaced_scalars[name] = (bwd_arr_name, bwd_desc)
         else:
             forward_sdfg.arrays[name].transient = False
 
+    for orig_name, (replaced_name, replaced_desc) in replaced_scalars.items():
+        del backward_input_arrays[orig_name]
+        backward_input_arrays[replaced_name] = replaced_desc
+
+    replaced_scalars = {}
+    for fwd_name, bwd_name in backward_result.required_grad_names.items():
+        desc = backward_sdfg.arrays[bwd_name]
+        if isinstance(desc, dt.Scalar):
+            arr_name, arr_desc = backward_sdfg.add_array(bwd_name + "_array",
+                                                         [1],
+                                                         desc.dtype,
+                                                         transient=False,
+                                                         storage=desc.storage,
+                                                         find_new_name=True)
+            desc.transient = True
+            bwd_copy_state = backward_sdfg.add_state_before(backward_state,
```

Comment:
add_state_before or after?

Reply:
Oh yeah, good catch. How did this work :confused: